### PR TITLE
Updates for LXD network API in 2.3

### PR DESCRIPTION
**Pull request by tych0**
Some thoughts:

we might not need the 3rd patch in the series (about falling back to the daily), but without it there's no way to get yakkety containers right now.

the second patch we definitely need, as it is a bug :)

### CODE REVIEW

File: tools/lxdclient/client_network.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+		n, err = client.NetworkGet(network.DefaultLXDBridge)
+		if err != nil {
+			return err
+		}
+	}
+
+	nicType := "macvlan"
+	if n.Type == "bridge" {
+		nicType = "bridged"
+	}
```
Comment (stgraber) [inline tools/lxdclient/client_network.go]: You could save yourself the second NetworkGet and that branch. Since you're calling NetworkCreate, you're guaranteed to be given a bridge.

The n.Type check is only used in our client for when someone does "lxc network attach eth0 blah" where eth0 isn't a bridge.

Comment (tych0) [inline tools/lxdclient/client_network.go]: That's only true as long as the default continues to be bridge, though.

Comment (stgraber) [inline tools/lxdclient/client_network.go]: Right, but changing network creation to something other than a bridge would be considered a major API breakage and would only happen if we rev the API to 2.0 (from 1.0).

Comment (tych0) [inline tools/lxdclient/client_network.go]: And this way if that happens we don't have to go back and patch juju :)

File: dependencies.tsv [hunk 2]
```diff
@@ -52,7 +52,7 @@ github.com/juju/xml	git	eb759a627588d35166bc505fceb51b88500e291e	2015-04-13T13:1
 github.com/juju/zip	git	f6b1e93fa2e29a1d7d49b566b2b51efb060c982a	2016-02-05T10:52:21Z
 github.com/julienschmidt/httprouter	git	77a895ad01ebc98a4dc95d8355bc825ce80a56f6	2015-10-13T22:55:20Z
 github.com/lunixbochs/vtclean	git	4fbf7632a2c6d3fbdb9931439bdbbeded02cbe36	2016-01-25T03:51:06Z
-github.com/lxc/lxd	git	0d172e3080f768fd419f0c97f5246983797db243	2016-08-12T05:08:18Z
```
Comment (dimitern) [inline dependencies.tsv]: Please, use `godeps github.com/juju/juju/... | grep lxc/lxd` to update that line, so it includes the correct date, not just the commit hash. Assuming your github.com/lxc/lxd work dir is at rev 48f.. it will produce the correct line to replace.

Comment (tych0) [inline dependencies.tsv]: Thanks, updated.

File: tools/lxdclient/client.go [hunk 3]
```diff
@@ -123,15 +124,24 @@ func Connect(cfg Config) (*Client, error) {
...
+			networkAPISupported = true
+		}
+	}
+
+	var bridgeName string
+	if remoteID == remoteIDForLocal {
+		// If this is the LXD provider on the localhost, let's do an extra check to
+		// make sure the default profile has a correctly configured bridge, and
+		// which one is it.
+		bridgeName, err = verifyDefaultProfileBridgeConfig(raw, networkAPISupported)
```
Comment (wallyworld) [inline tools/lxdclient/client.go]: We are ignoring the error here, oops

File: container/lxd/initialisation.go [hunk 4]
```diff
@@ -140,6 +143,20 @@ var configureZFS = func() {
 }
 
 var configureLXDBridge = func() error {
+	client, err := ConnectLocal()
+	if err != nil {
+		return err
```
Comment (wallyworld) [inline container/lxd/initialisation.go]: For consistency, should be errors.Trace(err)

File: tools/lxdclient/client.go [hunk 5]
```diff
@@ -261,6 +272,17 @@ func verifyDefaultProfileBridgeConfig(client *lxd.Client) (string, error) {
 	// with it, so let's just use whatever they set up.
 	eth0, ok := config.Devices[configEth0]
 	if !ok {
+		/* on lxd >= 2.3, there is nothing in the default profile
+		 * w.r.t. eth0, because there is no lxdbr0 by default. Let's
+		 * handle this case and configure one now.
+		 */
+		if networkAPISupported {
+			if err := CreateDefaultBridgeInDefaultProfile(client); err != nil {
+				return "", errors.Errorf("couldn't create default bridge %v", err)
```
Comment (wallyworld) [inline tools/lxdclient/client.go]: Should be errors.Annotate()

File: tools/lxdclient/client.go [hunk 6]
```diff
@@ -261,6 +272,17 @@ func verifyDefaultProfileBridgeConfig(client *lxd.Client) (string, error) {
...
+			}
+
+			return network.DefaultLXDBridge, nil
+		}
 		return "", errors.Errorf("unexpected LXD %q profile config without eth0: %+v", defaultProfileName, config)
```
Comment (wallyworld) [inline tools/lxdclient/client.go]: errors.Annotatef() here
(and anywhere else relevant)

File: tools/lxdclient/client_image.go [hunk 7]
```diff
@@ -98,7 +98,7 @@ func (i *imageClient) EnsureImageExists(series string, sources []Remote, copyPro
 	// at private methods so we can't easily tweak it.
 	name := i.ImageNameForSeries(series)
 
-	var lastErr error
+	lastErr := errors.Errorf("image not imported!")
```
Comment (wallyworld) [inline tools/lxdclient/client_image.go]: errors.New()

File: tools/lxdclient/client_network.go [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+	NetworkCreate(name string, config map[string]string) error
+	NetworkGet(name string) (shared.NetworkConfig, error)
+}
+
+type networkClient struct {
+	raw rawNetworkClient
+	supported bool
+}
+
+func (c *networkClient) NetworkCreate(name string, config map[string]string) error {
```
Comment (wallyworld) [inline tools/lxdclient/client_network.go]: exported functions need to be commented, here and below

File: tools/lxdclient/client_network.go [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+	if !c.supported {
+		return fmt.Errorf("network API not supported on this remote")
```
Comment (wallyworld) [inline tools/lxdclient/client_network.go]: should be using juju/errors
and there's a NotSupported error type

File: tools/lxdclient/client.go [hunk 10]
```diff
@@ -140,6 +150,7 @@ func Connect(cfg Config) (*Client, error) {
 		profileClient:            &profileClient{raw},
 		instanceClient:           &instanceClient{raw, remoteID},
 		imageClient:              &imageClient{raw, connectToRaw},
+		networkClient:		  &networkClient{raw, networkAPISupported},
```
Comment (wallyworld) [inline tools/lxdclient/client.go]: this line contains tabs not spaces so will fail go fmt

File: tools/lxdclient/remote.go [hunk 11]
```diff
@@ -44,8 +44,16 @@ var CloudImagesRemote = Remote{
 	ServerPEMCert: "",
 }
 
+var CloudImagesDailyRemote = Remote{
```
Comment (wallyworld) [inline tools/lxdclient/remote.go]: Pre-existing code, but not of the exported consts or vars etc in this file are commented and they should be.
Would be great to see a driveby here to fix that

File: tools/lxdclient/client.go (hunk 6 above)
Comment (tych0) [inline tools/lxdclient/client.go]: In fact I don't think it applies here, because there's no error here to annotate.

**Comment by tych0** [comment]
In theory I've addressed all of these; thoughts welcome!

**Comment by tych0** [comment]
But also, I can't push the right button :(

File: tools/lxdclient/client.go (hunk 6 above)
Comment (wallyworld) [inline tools/lxdclient/client.go]: Doh, yes, sorry

File: tools/lxdclient/client_network.go [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+	return c.raw.NetworkGet(name)
+}
+
+type creator interface {
+	rawNetworkClient
+	ProfileDeviceAdd(profile, devname, devtype string, props []string) (*lxd.Response, error)
+	ProfileConfig(profile string) (*shared.ProfileConfig, error)
+}
+
+// Create a default bridge and (if necessary) insert it into the default profile.
```
Comment (wallyworld) [inline tools/lxdclient/client_network.go]: This is a nitpic sorry, but the Go doc convention is to start method comments with the name of the method.
// CreateDefaultBridgeInDefaultProfile creates a default bridge...

(here and elsewheren - the same also pplies to the comment on the CloudImagesDailyRemote var)

**Comment by mitechie** [comment]
Will get this reviewed and landed this week as part of getting to GA. I've got two folks out today, but should get eyes on it tomorrow.

**Comment by tych0** [comment]
Cool, I just pushed an update to the comments.

File: tools/lxdclient/client_network.go [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+func CreateDefaultBridgeInDefaultProfile(client creator) error {
```
Comment (tych0) [inline tools/lxdclient/client_network.go]: In the container type case yes definitely, as there is no "user" that's on the machine there. In the provider case you could make an argument that it's not needed, but then it's just another step that someone has to do to start using the LXD provider, so I think it's a better UX to JFDI.

Comment (tych0) [inline tools/lxdclient/client_network.go]: I guess I don't understand. This code *does* contact LXD to set up the bridge. What code is it duplicating?

Comment (dooferlad) [inline tools/lxdclient/client_network.go]: @stgraber wrote:
"The n.Type check is only used in our client for when someone does "lxc network attach eth0 blah" where eth0 isn't a bridge."

This implies that your client (I assume lxc) can do the bridge set up and that very similar code has been put into Juju.

This is all a bit academic though because unless LXD grows an API call to set up the bridge that both Juju and LXC can call, then we need to duplicate the code.

**Comment by dooferlad** [comment]
$$merge$$

**Comment by tych0** [comment]
Hmm. so it looks like merge failed because of an additional dependency; looking into it now. What's the story on adding new dependencies?

**Comment by tych0** [comment]
Ok, I think I've untangled this: https://github.com/lxc/lxd/pull/2456

when the above is merged, i'll push a new commit with the updated hash and we can merge this one, I think.

**Comment by tych0** [comment]
I believe this is updated and should hopefully be able to merge now :)

**Comment by tych0** [comment]
$$merge$$

**Comment by AlexisBruemmer** [comment]
$$merge$$

**Comment by sinzui** [comment]
Build failed: Tests failed
build url: foo

**Comment by sinzui** [comment]
$$merge$$
